src/iterators/chunked_read_ahead_iterator.rs

```diff
@@ -0,0 +1,393 @@
+/// iterator
+pub struct ChunkedReadAheadIterator<T: Send + 'static> {
+    /// The recieving object that recieves chunks of ``T``. TODO - make this a Vec<T> when adding
+    /// chunking.
+    receiver: Option<Receiver<Option<Vec<T>>>>,
+    /// The handle to the thread that was spawned to read ahead on the iterator.
+    join_handle: Option<JoinHandle<()>>,
+    /// The most recent chunk recieved as an iterator. Used to produce owned ``T`` objects from
+    /// the chunk
+    current_chunk: IntoIter<T>,
+}
+
+impl<T> ChunkedReadAheadIterator<T>
+where
+    T: Send + 'static,
+{
+    /// Creates a new ``Self`` from an existing iterator, and parameters concerning the size of
+    /// underlying buffer elements.
+    ///
+    /// # Panics
+    ///
+    /// - panics if the underlying iterator panics
+    /// - panics if sending final chunk after an internal panic fails
+    /// - panics if sending the None kill signal to the recieiver after an internal panic fails
+    /// - panics if the spawned thread fails to spawn
+    pub fn new<I>(mut inner: I, chunk_size: usize, num_chunk_buffer_size: usize) -> Self
+    where
+        I: Iterator<Item = T> + Send + 'static,
+    {
+        assert_ne!(chunk_size, 0, "Chunk size cannot be zero!");
+        assert_ne!(num_chunk_buffer_size, 0, "Number of buffered chunks cannot be zero!");
+
+        // Create a channel over which we can send our chunks of ``T``
+        let (sender, receiver) = sync_channel(num_chunk_buffer_size);
+
+        // Create our spawned thread, holding on to the resulting handle for downstream error
+        // management.
+        let join_handle: JoinHandle<()> = thread::Builder::new()
+            .name("chunked_read_ahead_thread".to_owned())
+            .spawn(move || {
+                loop {
+                    let mut chunk = Vec::with_capacity(chunk_size);
+                    for _ in 0..chunk_size {
+                        match catch_unwind(AssertUnwindSafe(|| inner.by_ref().next())) {
+                            Ok(Some(val)) => chunk.push(val),
+                            Ok(None) => break,
+                            Err(e) => {
+                                sender
+                                    .send(Some(chunk))
+                                    .expect("Error sending final chunk before internal panic");
+                                sender
+                                    .send(None)
+                                    .expect("Error sending final None before internal panic");
+                                panic!("{}", extract_info_from_catch_unwind_panic(e));
+                            }
+                        }
+                    }
+                    if chunk.is_empty() {
+                        break;
+                    }
+                    let send_result = sender.send(Some(chunk));
+                    if send_result.is_err() {
+                        // TODO - Logging?
```

Comment:
Wasn't sure if we wanted to log this if errors occurred here. Would be easy to do, but wanted your feedback.